### PR TITLE
List method for TopicsController

### PR DESCRIPTION
**Pull request by Ielay**
- [x] List method of TopicsController
- [x] Test on list method
- [x] Update TopicsMapper
- [x] Add ACL into TopicsController
- [x] Test on ACL
- [x] Test on possible exceptions

Issue [#325](https://github.com/malinink/easy-tests/issues/325)

### CODE REVIEW

**Comment by malinink** [comment]
@Ielay provide tests pls

**Comment by Ielay** [comment]
@malinink Seems to be ok now. Ready to review.
P.s. My local problem was I just forget about up the migrations :)

**Comment by Ielay** [comment]
@malinink Could you check if I wrong in passing the parameter to get request in TopicControllerTest ? Travis writes that he can't get the queryParam that is required to a list method. I can swear that I do things correctly (e.g. just like there: https://reversecoding.net/spring-mvc-requestparam-binding-request-parameters/).

**Comment by malinink** [comment]
@Ielay It looks like parameter is valid. 409 error code means that response does not match doc. There is some comments in travis log, you probably miss them.
```
validation: GET#/v1/topics#200
2018-05-13 00:19:22.366  WARN 3178 --- [           main] e.swagger.SwaggerValidationInterceptor   : Swagger validation: GET#/v1/topics#200 - The REST response is invalid: [Path '/0'] Object has missing required properties (["subject"]), [Path '/1'] Object has missing required properties (["subject"])
```
```
MockHttpServletResponse:
           Status = 409
    Error message = null
          Headers = {Content-Type=[application/json;charset=UTF-8]}
     Content type = application/json;charset=UTF-8
             Body = [{"id":1,"name":"Name1","subject":null},{"id":2,"name":"Name2","subject":null}]
    Forwarded URL = null
   Redirected URL = null
          Cookies = []
```

**Comment by Ielay** [comment]
@malinink Well, created Json seems to be correct and seems to be matching with specification.
image

**Comment by malinink** [comment]
@Ielay but in test, return data does not match schema : `[{"id":1,"name":"Name1","subject":null},{"id":2,"name":"Name2","subject":null}]`

**Comment by malinink** [comment]
If you still need help, tell me pls.

**Comment by Ielay** [comment]
@malinink Yea, I need it. I actually have no idea why it has nulls in there. When I turn out the creation and use correct string value inside the method it doesn't help.
image

**Comment by malinink** [comment]
```
MockHttpServletResponse:
           Status = 409
    Error message = null
          Headers = {Content-Type=[application/json;charset=UTF-8]}
     Content type = application/json;charset=UTF-8
             Body = [{"id":1,"name":"Name1","subject":null},{"id":2,"name":"Name2","subject":null}]
    Forwarded URL = null
   Redirected URL = null
          Cookies = []

java.lang.AssertionError: [id=1].subject
Expected: a JSON object
     got: null
 ; [id=2].subject
Expected: a JSON object
     got: null
```
Problem is that SubjectModelEmpty does not map into Identity. You should add rule for that.
In simple controller you pass topic with subject, in test there is SubjectModelEmpty.
But! Case with SubjectModelEmpty should work also.

**Comment by Ielay** [comment]
@malinink Ya, I see, thanks.

**Comment by Ielay** [comment]
@malinink Finally ready for review

**Comment by Ielay** [comment]
@malinink Pls mark my PR "ready to review". I afraid you can forget to check it without this label :)

**Comment by Ielay** [comment]
@malinink Still waiting for.

File: src/test/java/easytests/api/v1/controllers/TopicsControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                        .with(name, topicsModels.get(0).getName())
+                        .with(subject, new JsonSupport()
+                                .with(id, topicsModels.get(0).getSubject().getId())
+                        )
+                )
+                .with(new JsonSupport()
+                        .with(id, topicsModels.get(1).getId())
+                        .with(name, topicsModels.get(1).getName())
+                        .with(subject, new JsonSupport()
+                                .with(id, topicsModels.get(0).getSubject().getId())
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/TopicsControllerTest.java]: 1 ?

File: src/main/java/easytests/api/v1/controllers/TopicsController.java [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
+import easytests.api.v1.exceptions.ForbiddenException;
+import easytests.api.v1.exceptions.NotFoundException;
 import easytests.api.v1.mappers.TopicsMapper;
+import easytests.api.v1.models.Topic;
+import easytests.core.models.SubjectModelInterface;
+import easytests.core.models.TopicModelInterface;
 import easytests.core.options.builder.TopicsOptionsBuilderInterface;
+import easytests.core.services.SubjectsServiceInterface;
 import easytests.core.services.TopicsServiceInterface;
+import java.util.List;
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/TopicsController.java:12]: wrong order

File: src/main/java/easytests/api/v1/controllers/TopicsController.java [hunk 3]
```diff
@@ -14,21 +24,43 @@
...
+
+    @Autowired
+    protected SubjectsServiceInterface subjectsService;
 
     @Autowired
     protected TopicsServiceInterface topicsService;
 
     @Autowired
-    private TopicsOptionsBuilderInterface topicsOptions;
+    private TopicsOptionsBuilderInterface topicsOptionsBuilder;
```
Comment (malinink) [inline src/main/java/easytests/api/v1/controllers/TopicsController.java:36]: wrong order

File: src/test/java/easytests/api/v1/controllers/TopicsControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        IntStream.range(0, 2).forEach(idx -> {
+            final TopicModel topicModel = new TopicModel();
+            topicModel.map(this.topicsSupport.getEntityFixtureMock(idx));
+            topicsModels.add(topicModel);
+        });
+        when(this.topicsService.findBySubject(any(SubjectModelInterface.class))).thenReturn(topicsModels);
+
+        when(this.subjectsService.find(Integer.valueOf(subjectIdParamValue)))
+                .thenReturn(new SubjectModelEmpty(Integer.valueOf(subjectIdParamValue)));
+        when(this.topicsService.findBySubject(new SubjectModelEmpty(Integer.valueOf(subjectIdParamValue))))
```
Comment (malinink) [inline src/test/java/easytests/api/v1/controllers/TopicsControllerTest.java]: duplicate 75 row

File: src/main/java/easytests/api/v1/mappers/TopicsMapper.java [hunk 5]
```diff
@@ -8,4 +11,12 @@
  */
 @Service("TopicsMapperV1")
 public class TopicsMapper extends ModelMapper {
+    public TopicsMapper() {
+        super();
+        this.createTypeMap(TopicModel.class, Topic.class)
+            .addMappings(mapper -> mapper.when(
```
Comment (malinink) [inline src/main/java/easytests/api/v1/mappers/TopicsMapper.java]: when is not necessary here, remove it

**Review by malinink — CHANGES_REQUESTED** [review]

File: src/main/java/easytests/api/v1/controllers/TopicsController.java (hunk 3 above)
Comment (Ielay) [reply src/main/java/easytests/api/v1/controllers/TopicsController.java:36]: @malinink With what?
I made autowired fields in that order just like you did in ObjectsController, isn't it?

File: src/main/java/easytests/api/v1/controllers/TopicsController.java (hunk 2 above)
Comment (Ielay) [reply src/main/java/easytests/api/v1/controllers/TopicsController.java:12]: @malinink Isn't we decide on this after looking on the first character in the path? If so, there is alphabet order: "e", "j" and then "o".
P.s. Checkstyle feels good on this order

Comment (Ielay) [reply src/main/java/easytests/api/v1/controllers/TopicsController.java:12]: @malinink Oh... I guess we should look at the first character of the each path layer. Okay.
UPD: Try to swap java.util.List and java.util.stream.Collectors and checkstyle gonna curse me.

**Comment by Ielay** [comment]
@malinink Fixed all of non-order remarks. Ready for review.

File: src/main/java/easytests/api/v1/controllers/TopicsController.java (hunk 2 above)
Comment (malinink) [reply src/main/java/easytests/api/v1/controllers/TopicsController.java:12]: Hm.. My Idea tells me that order is wrong, checkstyle does not check test files.
Please look again.
Other seems to be correct!

Comment (Ielay) [reply src/main/java/easytests/api/v1/controllers/TopicsController.java:12]: @malinink Ну я правда не знаю, в чем дело. Ни travis, ни локальный checkstyle не видит ошибок.
image

UPD: Кроме того, в других PR на списки точно такой же порядок.

**Review by malinink — APPROVED** [review]

**Pull request merged**